palace/utils/geodata.cpp

```diff
@@ -25,38 +25,61 @@ using CVector3dMap = Eigen::Map<const Eigen::Vector3d>;
 namespace
 {
 
-// Floating point precision for mesh IO. This precision is important, make sure nothing is
-// lost!
-const auto MSH_FLT_PRECISION = std::numeric_limits<double>::max_digits10;
-
 // Load the serial mesh from disk.
 std::unique_ptr<mfem::Mesh> LoadMesh(const std::string &, bool);
 
 // Optionally reorder mesh elements based on MFEM's internal reordeing tools for improved
 // cache usage.
-void ReorderMesh(mfem::Mesh &);
+void ReorderMesh(mfem::Mesh &mesh);
 
 // Generate element-based mesh partitioning, using either a provided file or METIS.
-std::unique_ptr<int[]> GetMeshPartitioning(mfem::Mesh &, int, const std::string &);
+std::unique_ptr<int[]> GetMeshPartitioning(mfem::Mesh &, int, const std::string & = "");
 
 // Cleanup the provided serial mesh by removing unnecessary domain and elements, adding
 // boundary elements for material interfaces and exterior boundaries, and adding boundary
 // elements for subdomain interfaces.
-std::map<int, std::array<int, 2>> CheckMesh(std::unique_ptr<mfem::Mesh> &,
-                                            const std::unique_ptr<int[]> &, const IoData &,
-                                            bool, bool, bool);
+std::map<int, std::array<int, 2>> CheckMesh(mfem::Mesh &, const std::unique_ptr<int[]> &,
+                                            const IoData &, bool, bool, bool);
 
 // Given a serial mesh on the root processor and element partitioning, create a parallel
-// mesh oer the given communicator.
-std::unique_ptr<mfem::ParMesh> DistributeMesh(MPI_Comm, std::unique_ptr<mfem::Mesh> &,
-                                              std::unique_ptr<int[]> &,
-                                              const std::string &);
+// mesh over the given communicator.
+std::unique_ptr<mfem::ParMesh> DistributeMesh(MPI_Comm, const std::unique_ptr<mfem::Mesh> &,
+                                              const std::unique_ptr<int[]> & = nullptr,
+                                              const std::string & = "");
 
 // Get list of domain and boundary attribute markers used in configuration file for mesh
 // cleaning.
 void GetUsedAttributeMarkers(const IoData &, int, int, mfem::Array<int> &,
                              mfem::Array<int> &);
 
+// Apply a scaling factor to the mesh vertices and nodes.
+void ScaleMesh(mfem::Mesh &mesh, double L)
+{
+  for (int i = 0; i < mesh.GetNV(); i++)
+  {
+    double *v = mesh.GetVertex(i);
+    std::transform(v, v + mesh.SpaceDimension(), v, [L](double val) { return val * L; });
+  }
+  if (mesh.GetNodes())
+  {
+    *mesh.GetNodes() *= L;
+  }
+}
+
+// Simplified helper for describing the element types in a mesh
+struct ElementTypeInfo
+{
+  bool has_simplices;
+  bool has_tensors;
+  bool has_wedges;
+  bool has_pyramids;
+};
+ElementTypeInfo CheckElements(mfem::Mesh &mesh)
+{
+  auto meshgen = mesh.MeshGenerator();
+  return {bool(meshgen & 1), bool(meshgen & 2), bool(meshgen & 4), bool(meshgen & 8)};
```

Comment:
I believe these are local to the process and not reduced in parallel. I think you need a reduction based on how they are used below.